### PR TITLE
Remove FILEDosToUnixPathA conversion

### PR DESCRIPTION
**Pull request by janvorli**
All file paths passed to the coreclr PAL are processed by this function to convert backslash characters to forward ones. This is causing problems when users have directories with names containing backslashes. Such problems were recently reported by people.

This change removes the `FILEDosToUnixPathA` and fixes one PAL test and one coreclr test that was passing backslash separated paths.

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

**Comment by jkotas** [comment]
There are a few places in the runtime that have similar problem. Could you please scan and fix them as well?

For example, https://github.com/dotnet/runtime/blob/2f9418db6b724b46b1d2358ce55c088d160a87eb/src/coreclr/utilcode/splitpath.cpp#L111

**Comment by jkotas** [comment]
This needs to be tracked as a breaking change.

**Comment by AaronRobinsonMSFT** [comment]
@elinor-fung Isn't there an issue that is tracking this behavior?

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
NativeAOT is also emulating this:

https://github.com/dotnet/runtime/blob/1076cf52f60a845cf59e316345e2e599a6fcd3ac/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.NativeAot.Unix.cs#L14-L15

**Comment by janvorli** [comment]
@jkotas, @AaronRobinsonMSFT I have scanned the whole coreclr and nativeaot and modified all places where we were either converting the directory separators, or using both or even using backslashes on Unix too. And I've removed 
`_wsplitpath_s`.

File: src/coreclr/utilcode/longfilepathwrappers.cpp [hunk 1]
```diff
@@ -697,10 +697,10 @@ HRESULT LongFile::NormalizePath(SString & path)
 
 BOOL LongFile::ContainsDirectorySeparator(SString & path)
 {
-    return path.Find(path.Begin(), DirectorySeparatorChar) || path.Find(path.Begin(), AltDirectorySeparatorChar);
+    return path.Find(path.Begin(), DirectorySeparatorChar);
```
Comment (jkotas) [inline src/coreclr/utilcode/longfilepathwrappers.cpp:700]: Should `AltDirectorySeparatorChar` handling stay here for Windows?

File: src/coreclr/utilcode/longfilepathwrappers.cpp [hunk 2]
```diff
@@ -534,8 +534,8 @@ CopyFileExWrapper(
 
 //Implementation of LongFile Helpers
 const WCHAR LongFile::DirectorySeparatorChar = W('\\');
-const WCHAR LongFile::AltDirectorySeparatorChar = W('/');
 #ifdef HOST_WINDOWS
+const WCHAR LongFile::AltDirectorySeparatorChar = W('/');
```
Comment (jkotas) [inline src/coreclr/utilcode/longfilepathwrappers.cpp:538]: Why are we compiling any of this for Unix? Should this file be Windows-specific?

File: src/coreclr/vm/peimage.cpp [hunk 3]
```diff
@@ -593,7 +594,7 @@ void PEImage::EnumMemoryRegions(CLRDataEnumMemoryFlags flags)
                         pCvInfo->path[MAX_LONGPATH - 1] = '\0';
 
                         //Find the filename from pdb full path
-                        char* fileName = strrchr(pCvInfo->path, '\\');
+                        char* fileName = strrchr(pCvInfo->path, DIRECTORY_SEPARATOR_CHAR_A);
```
Comment (jkotas) [inline src/coreclr/vm/peimage.cpp:597]: This is searching in a path embedded in .pdb file. I am not sure whether this change is right.

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/peimage.cpp:597]: If we embed on non-Windows, I would assume the paths could have `/`.

File: src/coreclr/utilcode/longfilepathwrappers.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/utilcode/longfilepathwrappers.cpp:700]: This function is actually unused, I will remove it. However, the `IsDirectorySeparator` where I've also removed the alt directory handling needs to get it back, as it is used on the code path of the `LoadLibraryExWrapper` where the normalization of the path was added few years ago to handle issue with some NuGet packages containing paths with all forward slashes.

Comment (janvorli) [reply src/coreclr/utilcode/longfilepathwrappers.cpp:700]: I am not sure though with all the usages of the `NormalizePath` that calls the `IsDirectorySeparator`. All the wrappers in this file that get file path as an input call the `NormalizePath`. @jkotas do you see a reason for accepting both forms of slash in methods like `GetFileAttributesWrapper`, `CreateFileWrapper` etc? I have thought the `LoadLibraryExWrapper` is an exception.

File: src/coreclr/utilcode/longfilepathwrappers.cpp (hunk 2 above)
Comment (janvorli) [reply src/coreclr/utilcode/longfilepathwrappers.cpp:538]: This file contains also wrappers for functions that return string in a buffer and those wrappers get SString argument. But I will ifdef out the functions that are not this case and the LongFile class.

File: src/coreclr/vm/peimage.cpp (hunk 3 above)
Comment (janvorli) [reply src/coreclr/vm/peimage.cpp:597]: It looks like that for managed assemblies we generate PDB files in "BSJB" format. Only native Windows libraries have the RSDS format. This code path is taken only for the latter (there is an "if" a bit above). So this change needs to be reserved and we would always get files with backslash separators here.

File: src/coreclr/utilcode/longfilepathwrappers.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/utilcode/longfilepathwrappers.cpp:700]: Do Windows APIs consider the slash and backslash to be directory separators? That would be surprising to me. I don't see that mentioned in the MSDN doc at https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file.

Comment (janvorli) [reply src/coreclr/utilcode/longfilepathwrappers.cpp:700]: Hmm, I've tried it and they really do support that. Interesting.

**Comment by janvorli** [comment]
The `AppHost.Bundle.Tests.BundleExtractToSpecificPath.Bundle_Extraction_To_Relative_Path_Succeeds` fails with these changes for this `[InlineData("foo\\bar", BundleOptions.BundleNativeBinaries)]`. This test case should be Windows specific after this change as it expects the foo and bar to be separate directories. We already have an exclusion check for similar case with forward slash on Windows, so I will mimic that.

**Comment by janvorli** [comment]
@jkotas, @AaronRobinsonMSFT the CI is green and I believe I have addressed all of the feedback. I'll merge it in the morning unless you have an additional feedback.

**Comment by janvorli** [comment]
The breaking change issue is here: https://github.com/dotnet/docs/issues/32906

**Pull request merged**